sycl/doc/extensions/FreeFunctionQueries/SYCL_INTEL_free_function_queries.asciidoc

```diff
@@ -0,0 +1,180 @@
+= SYCL_INTEL_global_work_items
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+
+
+:blank: pass:[ +]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+== Introduction
+IMPORTANT: This specification is a draft.
+
+NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by permission by Khronos.
+
+NOTE: This document is better viewed when rendered as html with asciidoctor.  GitHub does not render image icons.
+
+This document describes an extension that adds features for SYCL work items and groups to be available globally.
+
+
+== Name Strings
+
++SYCL_INTEL_global_work_items+
+
+== Notice
+
+Copyright (c) 2020 Intel Corporation.  All rights reserved.
+
+== Status
+
+Working Draft
+
+This is a preview extension specification, intended to provide early access to a feature for review and community feedback. When the feature matures, this specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are subject to change they are not intended to be used by shipping software products.
+
+== Version
+
+Built On: {docdate} +
+Revision: 1
+
+== Contact
+Ruslan Arutyunyan, Intel (ruslan 'dot' arutyunyan 'at' intel 'dot' com)
+
+John Pennycook, Intel (john 'dot' pennycook 'at' intel 'dot' com)
+
+== Dependencies
+
+This extension is written against the SYCL 2020 Provisional specification, Revision 1.
+
+== Overview
+
+The extension allows to get `item`, `nd_item`, `group`, `sub_group` instances globally
+
+== Modifications of SYCL 2020 provisional specification
+
+=== Changes in 4.10.1.4 (item class)
+
+==== Changes in synopsis
+
+===== Add the following free function
+
+[source,c++,multiptr,linenums]
+----
+template <int dimensions, bool with_offset>
+item<dimensions> this_item();
+----
+
+===== Add the table with free functions description
+
+|===
+a|
+[source,c++,multiptr,linenums]
+----
+template <int dimensions, bool with_offset>
+item<dimensions> this_item();
+---- a|
+Returns the `item` instance representing the current work item in a `range`. The prerequisites:
+
+* `dimensions` value passed to `this_item` function shall match `dimensions` value passed to the `range` when the `parallel_for` is called
```

Comment:
Do we need to add a prerequisite that `with_offset` can only be false if the kernel was started with no offset?

Reply:
I don't think so, it's always safe to construct `item<Dims, true>` from `item<Dims, false>`.

Reply:
But is it always safe to construct `item<Dims, false>`?

Reply:
I think it should be. On the device you just don't read the offset with the SPIR-V builtin.

On the host you need to do additional step to achieve that. Basically, even if you call the `parallel_for` with offset, on the implementation level you need to store both `item<Dims, true>` and `item<Dims, false>`

Reply:
Deeper investigation shows that it's not always safe. Updated the extension as we agreed with John

Reply:
Agreed with Roland on the other solution. Just remove the `with_offset` template parameter and return `item` where `with_offset` is always `true`. It simplifies the usage of that API.